# DataFrames Exercises

Do your work for this exercise in a python script or a jupyter notebook with the name dataframes.py or dataframes.ipynb.

For several of the following exercises, you'll need to load several datasets using the pydataset library. (If you get an error when trying to run the import below, use pip to install the pydataset package.)


            from pydataset import data


When the instructions say to load a dataset, you can pass the name of the dataset as a string to the data function to load the dataset. You can also view the documentation for the data set by passing the show_doc keyword argument.


       # data('mpg', show_doc=True) # view the documentation for the dataset
    mpg = data('mpg') # load the dataset and store it in a variable
    
    
All the datasets loaded from the pydataset library will be pandas dataframe

In [4]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

In [5]:
print(df)

       name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72


In [6]:
df.sample(3)

name  math  english  reading
7   Marie    93       63       90
11   Alan    92       62       72
1    Jane    88       79       67

In [7]:
english_passing = df[df.english > 70]

In [8]:
english_passing

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
9   Richard    69       80       94
10    Isaac    92       99       93

1). Copy the code from the lesson to create a dataframe full of student grades.

a). Create a column named passing_english that indicates whether each student has a passing grade in english.

In [9]:
df['passing_english'] = df.english > 70


In [10]:
df = df.drop(columns=['passing_grade'])

KeyError: "['passing_grade'] not found in axis"

b). Sort the english grades by the passing_english column. How are duplicates handled?



In [11]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

c). Sort the english grades first by passing_english and then by student name. All the students 
that are failing english should be first, and within the students that are failing english 
they should be ordered alphabetically. The same should be true for the students passing 
english. (Hint: you can pass a list to the .sort_values method)


In [12]:
df.sort_values(by='passing_english').sort_values(by='name')

name  math  english  reading  passing_english
4       Ada    77       92       98             True
11     Alan    92       62       72            False
8    Albert    92       62       87            False
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
7     Marie    93       63       90            False
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81            False

d). Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.



In [13]:
df.sort_values(by=['passing_english']).sort_values(by=['english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

e). Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.



In [14]:
df['overall_grade'] = df.mean(axis = 1)

/var/folders/b2/3fdswr7x4v5fhdt54wpcfmdw0000gn/T/ipykernel_933/628609030.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df['overall_grade'] = df.mean(axis = 1)


In [15]:
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True          57.00
1      Jane    88       79       67             True          58.75
2     Suzie    94       74       95             True          66.00
3     Billy    98       96       88             True          70.75
4       Ada    77       92       98             True          67.00
5      John    79       76       93             True          62.25
6    Thomas    82       64       81            False          56.75
7     Marie    93       63       90            False          61.50
8    Albert    92       62       87            False          60.25
9   Richard    69       80       94             True          61.00
10    Isaac    92       99       93             True          71.25
11     Alan    92       62       72            False          56.50

# 2. Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:



In [158]:
from pydataset import data
mpg = data('mpg')

In [17]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

    --> How many rows and columns are there?
    
    Answer: 234 rows × 11 columns

In [18]:
mpg.info

<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

    --> What are the data types of each column?

In [19]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

    --> Summarize the dataframe with .info and .describe

In [20]:
mpg.info


<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

In [21]:
mpg.describe

<bound method NDFrame.describe of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    

    --> Rename the cty column to city.

In [22]:
mpg = mpg.rename(columns={'cty':'city'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

    --> Rename the hwy column to highway.

In [23]:
mpg = mpg.rename(columns={'hwy':'highway'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

    --> Do any cars have better city mileage than highway mileage?
    
    Answer: 0

In [24]:
mpg[mpg.city > mpg.highway]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

    --> Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [25]:
mpg['mileage_difference'] = np.where((mpg['city'] >= mpg['highway']) & (mpg['city'] <= mpg['highway']), True, False)

mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact               False  
2    compact               False  
3    compact               False  
4    compact               False  
5    compact               False  
..       ...                 ...  
230  midsize               False  
231  midsize               False  
232  midsize               False  
233  midsize               False  
234  midsize               False  

[234 rows x 12 columns]

In [26]:
conditions = [mpg['city'] > mpg['highway'], mpg['city'] < mpg['highway'], mpg['city'] == mpg['highway']]

choices = ['city > highway', 'highway > city', 'city == highway']

mpg['mileage_difference'] = np.select(conditions, choices)

mpg


manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class mileage_difference  
1    compact     highway > city  
2    compact     highway > city  
3    compact     highway > city  
4    compact     highway > city  
5    compact     highway > city  
..       ...                ...  
230  midsize     highway > city  
231  midsize     highway > city  
232  midsize     highway > city  
233  midsize     highway > city  
234  midsize     highway > city  

[234 rows x 12 columns]

    --> Which car (or cars) has the highest mileage difference?

In [79]:
mpg[mpg.highest_mileage == mpg.highest_mileage.max()] 

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   

    fl       class mileage_difference  max_mileage_difference  highest_mileage  
107  c  subcompact     highway > city                      12               12  
223  d  subcompact     highway > city                      12               12

In [71]:
# Removes duplicate columns
mpg = mpg.loc[:,~mpg.columns.duplicated()]

    --> Which compact class car has the lowest highway mileage? The best?

In [140]:

compact_cars = mpg[mpg['class'] == 'compact']

low_mileage_compact = compact_cars[compact_cars.highway == compact_cars.highway.min()]

low_mileage_compact

manufacturer  model  displ  year  cyl     trans drv  city  highway fl  \
220   volkswagen  jetta    2.8  1999    6  auto(l4)   f    16       23  r   

       class mileage_difference  max_mileage_difference  highest_mileage  
220  compact     highway > city                      12                7

    --> Create a column named average_mileage that is the mean of the city and highway mileage.

In [155]:
mean_city = mpg['city'].mean()

mean_highway = mpg['highway'].mean()

mpg['average_mileage'] = ((mean_highway + mean_city) / 2)

mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  max_mileage_difference  mileage_difference  \
1    compact      highway > city                      12                  11   
2    compact      highway > city                      12                   8   
3    compact      highway > city                      12                  11   
4    compact      highway > city                      12                   9   
5    compact      highway > city                      12                  10   
..       ...                 ...                     ...                 ...   
230  midsize      highway > city                      12                   9   
231  midsize      highway > city                      12                   8   
232  midsize      highway > city                      12                  10   
233  midsize      highway > city                      12                   8   
234  midsize      highway > city                      12                   9   

     average_mileage  
1          20.149573  
2          20.149573  
3          20.149573  
4          20.149573  
5          20.149573  
..               ...  
230        20.149573  
231        20.149573  
232        20.149573  
233        20.149573  
234        20.149573  

[234 rows x 15 columns]

    --> Which dodge car has the best average mileage? The worst?

In [156]:
dodge_cars = mpg[mpg['manufacturer'] == 'dodge']

dodge_cars

dodge_car_best_average = dodge_cars[dodge_cars.average_mileage == dodge_cars.average_mileage.max()]

dodge_car_best_average


manufacturer                model  displ  year  cyl       trans drv  city  \
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   

    highway fl    class  mileage_difference  max_mileage_difference  \
38       24  r  minivan      highway > city                      12   
39       24  r  minivan      highway > city                      12   
40       22  r  minivan      highway > city                      12   
41       22  r  minivan      highway > city                      12   
42       24  r  minivan      highway > city                      12   
43       24  r  minivan      highway > city                      12   
44       17  e  minivan      highway > city                      12   
45       22  r  minivan      highway > city                      12   
46       21  r  minivan      highway > city                      12   
47       23  r  minivan      highway > city                      12   
48       23  r  minivan      highway > city                      12   
49       19  r   pickup      highway > city                      12   
50       18  r   pickup      highway > 

In [157]:
# Worst Dodge car average mileage
dodge_car_worst_average = dodge_cars[dodge_cars.average_mileage == dodge_cars.average_mileage.min()]
dodge_car_worst_average

manufacturer                model  displ  year  cyl       trans drv  city  \
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   

    highway fl    class  mileage_difference  max_mileage_difference  \
38       24  r  minivan      highway > city                      12   
39       24  r  minivan      highway > city                      12   
40       22  r  minivan      highway > city                      12   
41       22  r  minivan      highway > city                      12   
42       24  r  minivan      highway > city                      12   
43       24  r  minivan      highway > city                      12   
44       17  e  minivan      highway > city                      12   
45       22  r  minivan      highway > city                      12   
46       21  r  minivan      highway > city                      12   
47       23  r  minivan      highway > city                      12   
48       23  r  minivan      highway > city                      12   
49       19  r   pickup      highway > city                      12   
50       18  r   pickup      highway > 

# 3. Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

    --> How many rows and columns are there?

In [160]:
mammals = data('Mammals')
mammals

weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    130.000   70.0    False     False
25    120.000   80.0    False     False
26    120.000   61.0    False     False
27    110.000   33.0    False     False
28    100.000   64.0    False     False
29     85.000   55.0    False     False
30     80.000   65.0    False     False
31     72.000   56.0    False     False
32     70.000   45.0    False     False
33     65.000   60.0    False     False
34     62.000   81.0    False     False
35     50.000  100.0    False     False
36     50.000   60.0    False     False
37     50.000   40.0    False     False
38     50.000   47.0    False     False
39     37.000  105.0    False     False
40     35.000   80.0    False     False
41     34.000   97.0    False     False
42     30.000   97.0    False     False
43     30.000   80.0    False     False
44     30.000   45.0    False     False
45     20.000   81.0    False     False
46    400.000   40.0    False     False
47    300.000   48.0    False     False
48    230.000   56.0    False     False
49    150.000   59.0    False     False
50    135.000   48.0    False     False
51     65.000   65.0    False     False
52     60.000   60.0    False     False
53     55.000  110.0    False     False
54     45.000   50.0    False     False
55     40.000   64.0    False     False
56     25.000   67.0    False     False
57     20.000   70.0    False     False
58     16.000   65.0    False     False
59     12.000   24.0    False      True
60     11.000   30.0    False      True
61     10.000   56.0    False     False
62      7.000   60.0    False     False
63      6.000   72.0    False     False
64      5.000   64.0    False     False
65      5.000   27.0    False      True
66      3.000   16.0    False      True
67    127.000   32.0    False     False
68     70.000   40.0    False      True
69     13.000   37.0    False      True
70      9.000    3.2    False      True
71      4.000   16.0    False     False
72      0.600   36.0    False     False
73      0.600   20.0    False     False
74      0.550   27.0    False     False
75      0.500   18.0    False     False
76      0.400   20.0    False     False
77      0.300   13.0    False     False
78      0.250    9.7    False     False
79      0.220   15.0    False     False
80      0.110    9.0    False     False
81      0.100   17.0    False     False
82      0.056   21.0     True     False
83      0.050   11.0    False     False
84      0.045   16.0    False     False
85      0.035   32.0     True     False
86      0.035   14.0     True     False
87      0.030    6.8    False     False
88      0.030    9.1    False     False
89      0.025   11.0    False     False
90      0.025    8.6    False     False
91      0.018    8.9    False     False
92      0.016   13.0    False     False
93      0.100    4.0    False     False
94      0.100    2.4    False     False
95      0.016    3.6    False     False
96      4.600   64.0     True     False
97      4.400   72.0     True     False
98      4.000   72.0     True     False
99      3.500   56.0     True     False
100    

    --> What are the data types?

In [161]:
mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

    --> Summarize the dataframe with .info and .describe

In [162]:
mammals.info

<bound method DataFrame.info of        weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    13

    --> What is the the weight of the fastest animal?

In [165]:

fastest_mammal = mammals[mammals.speed == mammals.speed.max()]


fastest_mammal_weight = fastest_mammal.weight

fastest_mammal_weight

53    55.0
Name: weight, dtype: float64

    --> What is the overal percentage of specials?

In [168]:
mammals

weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    130.000   70.0    False     False
25    120.000   80.0    False     False
26    120.000   61.0    False     False
27    110.000   33.0    False     False
28    100.000   64.0    False     False
29     85.000   55.0    False     False
30     80.000   65.0    False     False
31     72.000   56.0    False     False
32     70.000   45.0    False     False
33     65.000   60.0    False     False
34     62.000   81.0    False     False
35     50.000  100.0    False     False
36     50.000   60.0    False     False
37     50.000   40.0    False     False
38     50.000   47.0    False     False
39     37.000  105.0    False     False
40     35.000   80.0    False     False
41     34.000   97.0    False     False
42     30.000   97.0    False     False
43     30.000   80.0    False     False
44     30.000   45.0    False     False
45     20.000   81.0    False     False
46    400.000   40.0    False     False
47    300.000   48.0    False     False
48    230.000   56.0    False     False
49    150.000   59.0    False     False
50    135.000   48.0    False     False
51     65.000   65.0    False     False
52     60.000   60.0    False     False
53     55.000  110.0    False     False
54     45.000   50.0    False     False
55     40.000   64.0    False     False
56     25.000   67.0    False     False
57     20.000   70.0    False     False
58     16.000   65.0    False     False
59     12.000   24.0    False      True
60     11.000   30.0    False      True
61     10.000   56.0    False     False
62      7.000   60.0    False     False
63      6.000   72.0    False     False
64      5.000   64.0    False     False
65      5.000   27.0    False      True
66      3.000   16.0    False      True
67    127.000   32.0    False     False
68     70.000   40.0    False      True
69     13.000   37.0    False      True
70      9.000    3.2    False      True
71      4.000   16.0    False     False
72      0.600   36.0    False     False
73      0.600   20.0    False     False
74      0.550   27.0    False     False
75      0.500   18.0    False     False
76      0.400   20.0    False     False
77      0.300   13.0    False     False
78      0.250    9.7    False     False
79      0.220   15.0    False     False
80      0.110    9.0    False     False
81      0.100   17.0    False     False
82      0.056   21.0     True     False
83      0.050   11.0    False     False
84      0.045   16.0    False     False
85      0.035   32.0     True     False
86      0.035   14.0     True     False
87      0.030    6.8    False     False
88      0.030    9.1    False     False
89      0.025   11.0    False     False
90      0.025    8.6    False     False
91      0.018    8.9    False     False
92      0.016   13.0    False     False
93      0.100    4.0    False     False
94      0.100    2.4    False     False
95      0.016    3.6    False     False
96      4.600   64.0     True     False
97      4.400   72.0     True     False
98      4.000   72.0     True     False
99      3.500   56.0     True     False
100    

    --> How many animals are hoppers that are above the median speed? What percentage is this?
    
    Answer: 7 above average speed
    

In [202]:
median_speed = mammals['speed'].median()
median_speed

hoppers = mammals[mammals.hoppers == True]

hopper_speed = hoppers[hoppers.speed > median_speed]

hopper_speed


weight  speed  hoppers  specials
96      4.6   64.0     True     False
97      4.4   72.0     True     False
98      4.0   72.0     True     False
99      3.5   56.0     True     False
100     2.0   64.0     True     False
101     1.9   56.0     True     False
102     1.5   50.0     True     False

In [189]:
hopper_mammals_avg

0.102803738317757

In [192]:
median_speed

48.0

In [194]:
hopper_mammals

1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
30     False
31     False
32     False
33     False
34     False
35     False
36     False
37     False
38     False
39     False
40     False
41     False
42     False
43     False
44     False
45     False
46     False
47     False
48     False
49     False
50     False
51     False
52     False
53     False
54     False
55     False
56     False
57     False
58     False
59     False
60     False
61     False
62     False
63     False
64     False
65     False
66     False
67     False
68     False
69     False
70     False
71     False
72     False
73     False
74     False
75     False
76     False
77     False

In [197]:
hoppers

weight  speed  hoppers  specials
82    0.056   21.0     True     False
85    0.035   32.0     True     False
86    0.035   14.0     True     False
96    4.600   64.0     True     False
97    4.400   72.0     True     False
98    4.000   72.0     True     False
99    3.500   56.0     True     False
100   2.000   64.0     True     False
101   1.900   56.0     True     False
102   1.500   50.0     True     False
103   1.500   40.0     True     False